File: rust/src/no349.rs
use std::collections::HashSet;

struct Solution;

impl Solution {
    #[allow(dead_code)]
    pub fn intersection(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        let set2: HashSet<i32> = nums2.into_iter().collect();
        nums1
            .into_iter()
            .filter(|item| set2.contains(item))
            .collect::<HashSet<i32>>()
            .into_iter()
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use crate::no349::Solution;

    #[test]
    fn test_intersection() {
        let nums1 = vec![1, 2, 2, 5];
        let nums2 = vec![2, 2];
        let vec_expect = vec![2];
        assert_eq!(vec_expect, Solution::intersection(nums1, nums2));
    }
}

File: java/src/no349.java
import org.junit.jupiter.api.Test;

import java.util.HashSet;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;

public class no349 {

    @Test
    public void test_intersect() {
        int[] nums1 = {1, 2, 2, 5};
        int[] nums2 = {2, 2};
        int[] expect = {2};
        int[] res = new no349().intersect(nums1, nums2);
        assertArrayEquals(res, expect);
    }

    public int[] intersect(int[] nums1, int[] nums2) {
        Set<Integer> set = new HashSet<>();
        for (int item : nums2) {
            set.add(item);
        }
        Set<Integer> res_set = new HashSet<>();
        for (int item : nums1) {
            if (set.contains(item)) {
                res_set.add(item);
            }
        }
        int[] res = new int[res_set.size()];
        int index = 0;
        for (int item : res_set) {
            res[index++] = item;
        }
        return res;
    }
}
